tests/compiler/test_exception.py

```diff
@@ -103,3 +107,81 @@ def test_optional_value_exception(snippetcompiler):
         "Optional variable accessed that has no value (attribute `n` of `__config__::Test (instantiated at {dir}/main.cf:13)`)"
         " (reported in self.n ({dir}/main.cf:8))",
     )
+
+
+def test_dataflow_exception(snippetcompiler):
+    snippetcompiler.setup_for_snippet(
+        """
+x = 0
+x = 1
+        """,
+    )
+    Config.set("compiler", "datatrace_enable", "true")
+    try:
+        compiler.do_compile()
+    except DoubleSetException as e:
+        assert e.msg.strip() == (
+            """
+value set twice:
+	old value: 0
+		set at {dir}/main.cf:2
+	new value: 1
+		set at {dir}/main.cf:3
+
+data trace:
+x
+├── 0
+│   SET BY `x = 0`
+│   AT {dir}/main.cf:2
+└── 1
+    SET BY `x = 1`
+    AT {dir}/main.cf:3
+            """.strip().format(  # noqa: W191, E101
+                dir=snippetcompiler.project_dir
+            )
+        )
+
+
+def test_dataflow_multi_exception(snippetcompiler):
+    snippetcompiler.setup_for_snippet(
+        """
+entity A:
+    number n
+end
+
+implement A using std::none
+
+x = A()
+y = A()
+
+x.n = y.n
+y.n = nn
+
+nn = mm
```

Comment:
This raises the question: do we consider `y.n` to the the root cause or the equivalence `[nn,mm]`

Reply:
i.e.  we can now go beyond checking only objects and to it into local variables as well.

Reply:
True, I would say the equivalence is the actual root cause. In the current implementation however, `y.n` will be reported because it's the root cause among those in the raised exception set. From there the data trace should show that the equivalence is the actual culprit.
Would you prefer the error message indicates the equivalence directly? In that case we could change the `root_causes` method to not return a subset of its input but instead the set of all its leaves in the rule resolution tree. The most challenging part would be cleanly updating the errors. This would probably require a new exception type to be defined as well.

Reply:
That would be the best solution, either make into another ticket or pack it into this one, that I leave up to you.

Reply:
I'm going to make another ticket. That way #1960 can be set as a prerequisite, I think it makes sense to decentralize the error handling first. Should this new ticket go into #1831 or #1830?

Reply:
Put it with the improvements